### PR TITLE
[NOT READY] Switch cucumber to js

### PR DESCRIPTION
**Pull request by TheMarex**
Tracking PR of @lbud works on switching the tests to cucumber-js.

### CODE REVIEW

**Comment by emiltin** [comment]
hi @lbud good to see you working on this. let me know if you have questions to the original ruby code - the code has grown a lot over time and could certainly use some good refactoring.

**Comment by emiltin** [comment]
how do you run the new js cucumber tests?

**Comment by lbud** [comment]
@emiltin thanks! Most of this was initially a direct port but now that I've become more familiar with the code I've been able to refactor out some repeated or unused code. There's still some refactoring that could be done but it's now in pretty good shape and for the sake of getting this up to speed with current `develop` and then getting open PRs working with JS-based cucumber tests, I'm going to revisit those a bit further down the line… 
Tests run only in node 4.0+. I've included a package.json so you should be able to run `npm install` and then

* `npm test` runs the `verify` profile
* `npm run cucumber features/` runs the default profile
* `npm run cucumber features/car/access.feature` to run individual tests

Remaining:
- [ ] rebase this to current `develop`
- [ ] update docs
- [ ] squash + merge

then visit changes introduced in https://github.com/Project-OSRM/osrm-backend/pull/1935 etc

cc @TheMarex

File: features/support/env.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+var path = require('path');
+var util = require('util');
+var fs = require('fs');
+var exec = require('child_process').execSync;
```
Comment (miccolis) [inline features/support/env.js]: We probably don't want the sync version as we hand this a callback.

File: features/support/exception_classes.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    toString () {
+        return util.format('*** %s\nLast %s from %s:\n%s\n', this.msg, this.lines, this.log, this.extract);
+    }
+
+    logTail (path, n) {
+        var expanded = path.resolve(this.TEST_FOLDER, path);
+        if (!fs.existsSync(expanded)) {
+            fs.readFile(expanded, (err, data) => {
```
Comment (miccolis) [inline features/support/exception_classes.js]: Mix of sync/async here is confusing.

File: features/step_definitions/requests.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    this.Then(/^response should be well-formed$/, () => {
+        this.ShouldBeWellFormed();
+    });
+
+    this.Then(/^status code should be (\d+)$/, (code, callback) => {
+        try {
+            this.json = JSON.parse(this.response.body);
+            assert.equal(this.json.status, parseInt(code));
+            callback();
```
Comment (miccolis) [inline features/step_definitions/requests.js]: minor thing - but the callback should probably be after the try/catch

File: features/step_definitions/routability.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                }
+
+                callback(null, r);
+            });
+        };
+
+        d3.queue()
+            .defer(testDirection, 'forw')
+            .defer(testDirection, 'backw')
+            .awaitAll((err, res) => {
```
Comment (miccolis) [inline features/step_definitions/routability.js]: Missing error handling, unless that is intentional.

File: package.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  "version": "0.0.0",
+  "private": true,
+  "description": "The Open Source Routing Machine is a high performance routing engine written in C++11 designed to run on OpenStreetMap data.",
+  "dependencies": {
+    "cucumber": "^0.9.4",
+    "d3-queue": "^2.0.3",
+    "node-timeout": "0.0.4",
+    "request": "^2.69.0",
+    "sha1": "^1.1.1",
+    "xmlbuilder": "^4.2.1"
```
Comment (miccolis) [inline package.json]: We are missing eslint.

File: features/support/env.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        });
+    };
+
+    this.verifyOSRMIsNotRunning = () => {
+        if (this.OSRMLoader.up()) {
+            throw new Error('*** osrm-routed is already running.');
+        }
+    };
+
+    this.verifyExistenceOfBinaries = () => {
```
Comment (miccolis) [inline features/support/env.js]: Should this method take a callback? It calls aync methods but doesn't wait for them to finish.

File: features/support/config.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
+var fs = require('fs');
+var path = require('path');
+var util = require('util');
+var sha1 = require('sha1');
```
Comment (miccolis) [inline features/support/config.js]: Can we use node's built in `crypto` module for this?

**Pull request merged**